### PR TITLE
Adds support for `serialize()` for Primitive, `RECORD`, `UNION` and `ARRAY` datatypes.

### PR DESCRIPTION
**Pull request by prashastia**
The implementation allows actual avro records to be serialized into storage api proto transferable protobufs.

The mapping is as follows:
1. Primitive Types - Object value is checked against the schema type passed and converted to its respective type. e.g If the schema indicates a `STRING` field, the value is validated to be a string and return the passed object as `STRING`.
2.  MAP - Not supported yet.
3. UNION - Validation is done to ensure the absence of multiple not null or only null type. In case of type of format ["null", datatype] the datatype value is converted. `null` values are allowed in this case.
4. ARRAY - Individual values are validated and converted one by one. In case of `null` value in array,  Error is thrown.
5. RECORD -  Recursive approach to convert each field items one by one. A `MESSAGE` is returned containing the respective fields.

/gcbrun

### CODE REVIEW

File: flink-connector-bigquery/pom.xml [hunk 1]
```diff
@@ -33,6 +33,12 @@ under the License.
     <packaging>jar</packaging>
 
     <dependencies>
+		<!-- joda dependency -->
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/pom.xml:36]: This should go under `utilities`: https://github.com/GoogleCloudDataproc/flink-bigquery-connector/blob/154aeb0f419a2aa5128df0eb91e05e1e9f86e581/flink-connector-bigquery/pom.xml#L72

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 2]
```diff
@@ -16,15 +16,202 @@
...
+    /**
+     * Initializer for the Serializer.
+     *
+     * @param bigQuerySchemaProvider BigQuerySchemaProvider for the Sink Table ({@link
+     *     BigQuerySchemaProvider} object)
+     */
+    public void init(BigQuerySchemaProvider bigQuerySchemaProvider) {
+        Preconditions.checkNotNull(
+                bigQuerySchemaProvider,
+                "BigQuerySchemaProvider not initialized before initializing Serializer.");
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:78]: Since this is a null check, a better error message would be: `BigQuerySchemaProvider not found while initializing Serializer`

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 3]
```diff
@@ -16,15 +16,202 @@
...
     public ByteString serialize(GenericRecord record) throws BigQuerySerializationException {
-        throw new UnsupportedOperationException("serialize method is not supported");
+        try {
+            return getDynamicMessageFromGenericRecord(record, this.descriptor).toByteString();
+        } catch (Exception e) {
+            throw new BigQuerySerializationException(e.getMessage(), e);
+        }
+    }
+
+    public Descriptor getDescriptor() {
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:91]: Why is this being exposed? Ideally BigQuerySchemaProvider should be used for this purpose.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 4]
```diff
@@ -16,15 +16,202 @@
...
+        Preconditions.checkNotNull(this.descriptor, "Descriptor not initialized!");
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:92]: This check should be performed in `init`. Also error message should be more elaborate. Something like `Destination BQ tables's proto schema not found`.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 2 above)
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/pom.xml (hunk 1 above)
Comment (prashastia) [reply flink-connector-bigquery/pom.xml]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 4 above)
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/pom.xml [hunk 5]
```diff
@@ -74,6 +74,12 @@ under the License.
             <groupId>dev.failsafe</groupId>
             <artifactId>failsafe</artifactId>
         </dependency>
+		<!-- joda dependency -->
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/pom.xml:77]: Nit: Remove this comment altogether.

File: flink-connector-bigquery/pom.xml [hunk 6]
```diff
@@ -74,6 +74,12 @@ under the License.
...
+		<dependency>
+			<groupId>joda-time</groupId>
+			<artifactId>joda-time</artifactId>
+			<version>2.10.10</version>
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/pom.xml:81]: As done in other dependencies, store version as pom property in root pom and use here.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 7]
```diff
@@ -16,15 +16,200 @@
...
+        PRIMITIVE_TYPE_ENCODERS.put(
+                Schema.Type.BYTES, o -> ByteString.copyFrom(((ByteBuffer) o).array()));
+    }
+
+    /**
+     * Initializer for the Serializer.
+     *
+     * @param bigQuerySchemaProvider BigQuerySchemaProvider for the Sink Table ({@link
+     *     BigQuerySchemaProvider} object)
+     */
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:74]:
```
    /**
     * Prepares the serializer before its serialize method can be called. It allows contextual
     * preprocessing after constructor and before serialize. The Sink will internally call this
     * method when initializing itself.
     *
     * @param bigQuerySchemaProvider {@link BigQuerySchemaProvider} for the destination table.
     */
```

Comment (jayehwhyehentee) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:74]: More importantly, this method needs to be defined in the parent `BigQueryProtoSerializer` interface and implemented here.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 8]
```diff
@@ -16,15 +16,200 @@
...
+                // Convert to Dynamic Message.
+                value = toProtoValue(fieldDescriptor, field.schema(), value);
+                builder.setField(fieldDescriptor, value);
+            }
+        }
+        return builder.build();
+    }
+
+    /**
+     * Function to convert a value of an AvroSchemaField value to required DynamicMessage value.
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:133]: Nit: Please link `{@link DynamicMessage}` when possible.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 9]
```diff
@@ -16,15 +16,200 @@
...
+                if (arrayElementType.isUnion()) {
+                    throw new IllegalArgumentException(
+                            "ARRAY cannot have multiple datatypes in BigQuery.");
+                }
+                // Convert each value one by one.
+                return StreamSupport.stream(iterable.spliterator(), false)
+                        .map(v -> toProtoValue(fieldDescriptor, arrayElementType, v))
+                        .collect(Collectors.toList());
+            case UNION:
+                Schema type = BigQuerySchemaProviderImpl.handleUnionSchema(avroSchema).getLeft();
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:167]: Static methods like `handleUnionSchema` should not be imported from implicit implementations. Better to define in some common location and import from there.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 10]
```diff
@@ -16,15 +16,200 @@
...
+     * @param value Value of the Avro Schema Field.
+     * @return Converted Object.
+     * @throws UnsupportedOperationException In case schema is of type MAP.
+     */
+    private static Object toProtoValue(
+            FieldDescriptor fieldDescriptor, Schema avroSchema, Object value)
+            throws UnsupportedOperationException {
+        switch (avroSchema.getType()) {
+            case RECORD:
+                // Recursion
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:147]: Very evident comments can be avoided.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 11]
```diff
@@ -16,15 +16,200 @@
...
+                        .map(v -> toProtoValue(fieldDescriptor, arrayElementType, v))
+                        .collect(Collectors.toList());
+            case UNION:
+                Schema type = BigQuerySchemaProviderImpl.handleUnionSchema(avroSchema).getLeft();
+                // Get the schema of the field.
+                return toProtoValue(fieldDescriptor, type, value);
+            case MAP:
+                throw new UnsupportedOperationException("Not supported yet");
+            default:
+                return scalarToProtoValue(avroSchema, value);
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:173]: Default behavior should be an error saying "unrecognized type". All known cases should be explicitly handled in the switch block.

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java [hunk 12]
```diff
@@ -16,20 +16,720 @@
...
+        BigQuerySchemaProvider bigQuerySchemaProvider =
+                TestBigQuerySchemas.getSchemaWithRequiredPrimitiveTypes();
+        AvroToProtoSerializer serializer = new AvroToProtoSerializer();
+        serializer.init(bigQuerySchemaProvider);
+        Descriptor descriptor = bigQuerySchemaProvider.getDescriptor();
+        Schema avroSchema = bigQuerySchemaProvider.getAvroSchema();
+
+        byte[] byteArray = "Any String you want".getBytes();
+        String recordSchemaString =
+                "{\"type\":\"record\",\"name\":\"required_record_field\",\"doc\":\"Translated Avro Schema for required_record_field\",\"fields\":[{\"name\":\"species\",\"type\":\"string\"}]}}";
```
Comment (jayehwhyehentee) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:58]: Better to split into multiple lines for readability.

Comment (prashastia) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 11 above)
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 10 above)
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 8 above)
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 13]
```diff
@@ -16,15 +16,200 @@
...
+     * @param fieldDescriptor {@link com.google.protobuf.Descriptors.FieldDescriptor} Object
+     *     describing the sink table field to which given value needs to be converted to.
+     * @param avroSchema {@link Schema} Object describing the value of Avro Schema Field.
+     * @param value Value of the Avro Schema Field.
+     * @return Converted Object.
+     * @throws UnsupportedOperationException In case schema is of type MAP.
+     */
+    private static Object toProtoValue(
+            FieldDescriptor fieldDescriptor, Schema avroSchema, Object value)
+            throws UnsupportedOperationException {
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 14]
```diff
@@ -16,15 +16,200 @@
...
+     * Function to convert a Generic Avro Record to Dynamic Message to write using the Storage Write
+     * API.
+     *
+     * @param element {@link GenericRecord} Object to convert to {@link DynamicMessage}
+     * @param descriptor {@link Descriptor} describing the schema of the sink table.
+     * @return {@link DynamicMessage} Object converted from the Generic Avro Record.
+     */
+    public static DynamicMessage getDynamicMessageFromGenericRecord(
+            GenericRecord element, Descriptor descriptor) {
+        Schema schema = element.getSchema();
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 7 above)
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Fixed.

File: flink-connector-bigquery/pom.xml (hunk 6 above)
Comment (prashastia) [reply flink-connector-bigquery/pom.xml]: Fixed.

File: flink-connector-bigquery/pom.xml (hunk 5 above)
Comment (prashastia) [reply flink-connector-bigquery/pom.xml]: Fixed.

File: pom.xml [hunk 15]
```diff
@@ -195,6 +195,11 @@ under the License.
                 <artifactId>failsafe</artifactId>
                 <version>3.3.2</version>
             </dependency>
+			<dependency>
+				<groupId>joda-time</groupId>
+				<artifactId>joda-time</artifactId>
+				<version>2.10.10</version>
+			</dependency>
```
Comment (jayehwhyehentee) [inline pom.xml:202]: Indentation seems off.

**Review by jayehwhyehentee — APPROVED** [review]
Thank you for working on this and addressing comments. Great PR :)

**Comment by jayehwhyehentee** [comment]
@clmccart Please take a look at this PR.

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java [hunk 16]
```diff
@@ -16,20 +16,731 @@
...
+
+        byte[] byteArray = "Any String you want".getBytes();
+        Schema recordSchema = avroSchema.getField("required_record_field").schema();
+        GenericRecord record =
+                new GenericRecordBuilder(avroSchema)
+                        .set("number", -7099548873856657385L)
+                        .set("price", 0.5616495161359795)
+                        .set("species", "String")
+                        .set("flighted", true)
+                        .set("sound", ByteBuffer.wrap(byteArray))
```
Comment (clmccart) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:64]: should this cover all the primitive types that we support? for example, what about enum?

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java [hunk 17]
```diff
@@ -16,20 +16,731 @@
...
+        assertEquals(
+                "hello",
+                message.getField(
+                        descriptor
+                                .findNestedTypeByName(fieldDescriptor.toProto().getTypeName())
+                                .findFieldByNumber(1)));
+    }
+
+    @Test
+    public void testRemainingPrimitiveTypesConversion() {
```
Comment (clmccart) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:91]: let's also add a test for unsupported primitive types (ensuring it throws the right error)

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java [hunk 18]
```diff
@@ -16,20 +16,731 @@
...
+                        .set("float_field", null)
+                        .set("enum_field", null)
+                        .build();
+        ByteString byteString = serializer.serialize(record);
+        assertEquals("", byteString.toStringUtf8());
+    }
+
+    // ------------ Test Schemas with Record of Different Types -----------
+    @Test
+    public void testRecordOfArrayConversation() {
```
Comment (clmccart) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:158]: here and below: conversion instead of conversation :)

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java
Comment (clmccart) [inline flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:1]: loving the test coverage here! but the structure of the tests are kind of hard to follow. a few things i think could help here:

1. organizing each test into "arrange", "act", and "assert" blocks so it is clear what is under test under what conditions
2. name the test following the "test<Scenario><Action><ExpectedBehavior>" schema (ie, testRecordHasValidPrimitiveTypesGetDynamicMessageFromGenericRecordConvertsFieldsCorrectly) -- note that this will make the test names much more verbose but also much more helpful
3. parameterize some of the tests (ie have a single test for getDynamicMessageFromGenericRecord that can run for both valid and invalid field types

no need to completely rewrite this file but if you could take another pass at it with the above things in mind and see if there's any low-hanging fruit, i think that would be super helpful. but definitely timebox it.

the only reason i bring it up at all is because often times people will look at tests to get an understanding of how things like conversion logic behaves so i think it's especially important for the tests to be clear since they effectively serve as a form of documentation

**Review by clmccart — CHANGES_REQUESTED** [review]

File: flink-connector-bigquery/pom.xml [hunk 19]
```diff
@@ -74,6 +74,10 @@ under the License.
             <groupId>dev.failsafe</groupId>
             <artifactId>failsafe</artifactId>
         </dependency>
+		<dependency>
+			<groupId>joda-time</groupId>
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/pom.xml:78]: Will be used for the upcoming PR, removing from here. Will add in the next one.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 20]
```diff
@@ -16,15 +16,253 @@
...
+        this.descriptor = derivedDescriptor;
+    }
+
     @Override
     public ByteString serialize(GenericRecord record) throws BigQuerySerializationException {
-        throw new UnsupportedOperationException("serialize method is not supported");
+        try {
+            return getDynamicMessageFromGenericRecord(record, this.descriptor).toByteString();
+        } catch (Exception e) {
+            throw new BigQuerySerializationException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:69]: Yep, kind of. Any type of exception thrown is thrown again as a [BigQuerySerializationException](https://github.com/GoogleCloudDataproc/flink-bigquery-connector/blob/154aeb0f419a2aa5128df0eb91e05e1e9f86e581/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/exceptions/BigQuerySerializationException.java#L20) with same error message. Fixed to hide the exception responsible and just use the `BigQuerySerializationException`.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 21]
```diff
@@ -16,15 +16,253 @@
...
+     *
+     * @param element {@link GenericRecord} Object to convert to {@link DynamicMessage}
+     * @param descriptor {@link Descriptor} describing the schema of the sink table.
+     * @return {@link DynamicMessage} Object converted from the Generic Avro Record.
+     */
+    public static DynamicMessage getDynamicMessageFromGenericRecord(
+            GenericRecord element, Descriptor descriptor) {
+        Schema recordSchema = element.getSchema();
+        DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+        // Get the record's schema and find the field descriptor for each field one by one.
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:85]: Yep, you are right. Fixed.

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java (hunk 18 above)
Comment (prashastia) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:158]: Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 22]
```diff
@@ -16,15 +16,253 @@
...
+                    TIMESTAMP-MILLIS - LONG
+                    UUID - STRING
+                    TIME-MILLIS - INT/STRING
+                    TIME-MICROS - LONG/STRING
+                    LOCAL-TIMESTAMP-MICROS - LONG/STRING
+                    LOCAL-TIMESTAMP-MILLIS - LONG/STRING
+                    GEOGRAPHY - STRING
+                    JSON - STRING
+                */
+            case LONG:
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:162]: same as `INT`, if there is a logical type it would be handled (logical type handling in the next [PR](https://github.com/GoogleCloudDataproc/flink-bigquery-connector/pull/101)), otherwise the value passed would be returned as it is.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 23]
```diff
@@ -16,15 +16,253 @@
...
+     * @return Converted Object.
+     */
+    private static Object toProtoValue(
+            FieldDescriptor fieldDescriptor, Schema avroSchema, Object value) {
+        switch (avroSchema.getType()) {
+            case RECORD:
+                return getDynamicMessageFromGenericRecord(
+                        (GenericRecord) value, fieldDescriptor.getMessageType());
+            case ARRAY:
+                Iterable<Object> iterable = (Iterable<Object>) value;
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:127]: ACK, Fixed.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 24]
```diff
@@ -16,15 +16,253 @@
...
+    public static DynamicMessage getDynamicMessageFromGenericRecord(
+            GenericRecord element, Descriptor descriptor) {
+        Schema recordSchema = element.getSchema();
+        DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+        // Get the record's schema and find the field descriptor for each field one by one.
+        for (Schema.Field field : recordSchema.getFields()) {
+            // In case no field descriptor exists for the field, throw an error as we have
+            // incompatible schemas.
+            FieldDescriptor fieldDescriptor =
+                    Preconditions.checkNotNull(
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:90]: All types of errors thrown will be caught later and enclosed as a `BigQuerySerializationException` with the same error message as that of exception thrown.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 25]
```diff
@@ -16,15 +16,253 @@
...
+                            descriptor.findFieldByName(field.name().toLowerCase()));
+            // Get the value for a field.
+            // Check if the value is null.
+            @Nullable Object value = element.get(field.name());
+            if (value == null) {
+                // If the field required, throw an error.
+                if (fieldDescriptor.isRequired()) {
+                    throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:98]: Same as this [comment](https://github.com/GoogleCloudDataproc/flink-bigquery-connector/pull/104/files#r1557339757)

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java (hunk 16 above)
Comment (prashastia) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:64]: Yep, we do test for all primitive types offered by avro. [Primitive Types](https://github.com/prashastia/flink-bigquery-connector/blob/7af9b7190ae53afb25e5f9c1aca2a610a5f40388/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/TestBigQuerySchemas.java#L77) tests for the Datatypes present in BQ, while [Remaining Primitive Type ](https://github.com/prashastia/flink-bigquery-connector/blob/7af9b7190ae53afb25e5f9c1aca2a610a5f40388/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/TestBigQuerySchemas.java#L103) checks for other primitive types like ENUM (not present in BQ)

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 26]
```diff
@@ -16,15 +16,253 @@
...
+            // incompatible schemas.
+            FieldDescriptor fieldDescriptor =
+                    Preconditions.checkNotNull(
+                            descriptor.findFieldByName(field.name().toLowerCase()));
+            // Get the value for a field.
+            // Check if the value is null.
+            @Nullable Object value = element.get(field.name());
+            if (value == null) {
+                // If the field required, throw an error.
+                if (fieldDescriptor.isRequired()) {
```
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:97]: For the cases where value == null but the fieldDescriptor is `REPEATED` or `NULLABLE` we do nothing (this fieldDescriptor is absent in the `DynamicMessage`, Storage API automatically adds it as `NULL` or `[]` in case of `REPEATED` fields).

File: flink-connector-bigquery/pom.xml (hunk 19 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/pom.xml]: great; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: great; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:90]: got it; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: great; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:98]: got it; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: that makes sense. that behavior would remain intact if you consolidate the nested if statement into one-line, right?

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: great; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java [hunk 27]
```diff
@@ -16,15 +16,253 @@
...
+                    throw new IllegalArgumentException("Array cannot have NULLABLE datatype");
+                }
+                if (arrayElementType.isUnion()) {
+                    throw new IllegalArgumentException(
+                            "ARRAY cannot have multiple datatypes in BigQuery.");
+                }
+                // Convert each value one by one.
+                return StreamSupport.stream(iterable.spliterator(), false)
+                        .map(v -> toProtoValue(fieldDescriptor, arrayElementType, v))
+                        .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: great; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:149]: i'm not very familiar with java switch statements but does the string and long case fall through to the int case or does it fall through to the default case? either way, let's add a comment

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java (hunk 16 above)
Comment (clmccart) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:64]: ah okay sorry missed those test; thanks! you can resolve this convo

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java (hunk 18 above)
Comment (clmccart) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java]: great; thanks! you can resolve this convo

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: Removing the nesting would cause the case of `value == null` and not `isRequired()` to fall under the else case which is not desired.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (prashastia) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java:149]: Added!

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java (hunk 17 above)
Comment (prashastia) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java]: Added!

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java
Comment (prashastia) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:1]: Attempted to add Javadoc for each of the tests, added blocks for arrange, act, assert and renamed tests.

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: okay you can resolve this

File: flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java]: thanks; you can resolve this!

**Review by clmccart — APPROVED** [review]

File: flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java
Comment (clmccart) [reply flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java:1]: much easier to follow now; thanks so much!
